common/network-common/src/main/java/org/apache/spark/network/protocol/MessageWithHeader.java

```diff
@@ -44,6 +45,14 @@
   private long totalBytesTransferred;
 
   /**
+   * When the write buffer size is larger than this limit, I/O will be done in chunks of this size.
+   * The size should not be too large as it will waste underlying memory copy. e.g. If network
+   * avaliable buffer is smaller than this limit, the data cannot be sent within one single write
+   * operation while it still will make memory copy with this size.
+   */
+  private static final int NIO_BUFFER_LIMIT = 512 * 1024;
```

Comment:
>Is it possible to know the accurate number? I guess not because it's OS dependent and may be changed vis OS settings.

There might be a way to get the accurate number of the network buffer, but I think it's meaningless to do that because even we get the accurate number, we cannot guarantee the network send buffer is empty each time we write the data, which means, it's always possible that we can only write part of the data whatever size we set `NIO_BUFFER_LIMIT`. We can only say the smaller the `NIO_BUFFER_LIMIT` is, the less redundant copy will be made.